### PR TITLE
instructions for building pony with MEMTRACK

### PR DESCRIPTION
**Pull request by kapilash**
Adding instructions for building pony with MEMTRACK functionality as suggested in the [PR 3863](https://github.com/ponylang/ponyc/pull/3863)

### CODE REVIEW

File: content/reference/index.md [hunk 1]
```diff
@@ -13,6 +13,9 @@ title = "Reference"
 
 > A collection of LLDB extensions for working with the Pony programming language.
 
+- [Making pony track memory usage]({{< relref "pony.md >}})
```
Comment (SeanTAllen) [inline content/reference/index.md:16]: this relref isn't correct. it should be to the other file you added.

Comment (SeanTAllen) [reply content/reference/index.md:16]: here and elsewhere in the new content it should be `Pony` not `pony`.

File: content/reference/pony-track-memory-usage.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
++++
+title = "Tracking memory usage at runtime"
++++
+
+## Tracking memory consumed by an actor
+The pony runtime can be built to track memory usage. This is helpful when trying to understand what components are using up memory.
+
+The pony runtime can be configured to be built with the following functions exposed by the pony runtime system:
```
Comment (SeanTAllen) [inline content/reference/pony-track-memory-usage.md:8]:
```suggestion
When configured, the following functions are exposed:
```

File: content/reference/pony-track-memory-usage.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+- `ponyint_sched_static_mem_size` - returns the static memory used by the scheduler subsystem.
+- `ponyint_sched_static_alloc_size` - returns the static memory allocated by the scheduler subsystem.
+- `ponyint_sched_total_memory_size` - returns the total memory used by a scheduler thread.
+- `ponyint_sched_total_alloc_size` - returns the total memory allocated by a scheduler thread.
+- `ponyint_actor_mem_size` - returns the size of the given actor.given an actor, this function returns the size of the actor.
+- `ponyint_actor_alloc_size`  - returns the size of the memory pool allocated to the given actor.
+- `ponyint_actor_total_mem_size` - returns the memory that is actively used by the runtime for the given actor. 
+- `ponyint_actor_total_alloc_size` -  returns the memory allocated by the given actor whether it is actively used or not.
```
Comment (SeanTAllen) [inline content/reference/pony-track-memory-usage.md:17]: I'd prefer to see this as a list something like

```
- foo

hello

- bar

bye
```

File: content/reference/pony-track-memory-usage.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
++++
+title = "Tracking memory usage at runtime"
++++
+
+## Tracking memory consumed by an actor
+The pony runtime can be built to track memory usage. This is helpful when trying to understand what components are using up memory.
```
Comment (SeanTAllen) [inline content/reference/pony-track-memory-usage.md:6]: Please have a space between `##` and the paragraph that follows it.

File: content/reference/pony-track-memory-usage.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+The above functions can be used like any other [c functions](https://tutorial.ponylang.io/c-ffi/calling-c.html). 
+For example, in order to keep track of memory of any actor, we need to add the following in the beginning of a file. 
+
+```
+use @ponyint_actor_total_mem_size[U64](a : Any tag)
+```
+
+Later, this function can be used to figure out the total memory used by this actor by using one of the following
+
+```
```
Comment (SeanTAllen) [inline content/reference/pony-track-memory-usage.md:29]:
```suggestion
```pony
```

File: content/reference/pony-track-memory-usage.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+- `ponyint_actor_mem_size` - returns the size of the given actor.given an actor, this function returns the size of the actor.
+- `ponyint_actor_alloc_size`  - returns the size of the memory pool allocated to the given actor.
+- `ponyint_actor_total_mem_size` - returns the memory that is actively used by the runtime for the given actor. 
+- `ponyint_actor_total_alloc_size` -  returns the memory allocated by the given actor whether it is actively used or not.
+
+### Usage
+The above functions can be used like any other [c functions](https://tutorial.ponylang.io/c-ffi/calling-c.html). 
+For example, in order to keep track of memory of any actor, we need to add the following in the beginning of a file. 
+
+```
```
Comment (SeanTAllen) [inline content/reference/pony-track-memory-usage.md:23]:
```suggestion
```pony
```

File: content/reference/pony-track-memory-usage.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+Later, this function can be used to figure out the total memory used by this actor by using one of the following
+
+```
+    let size = @ponyint_actor_total_mem_size(this)
+    env.out.print("my size = " + size.string())
+```
+or,
+
+```
+
```
Comment (SeanTAllen) [inline content/reference/pony-track-memory-usage.md:36]: extra line can be deleted.

File: content/reference/pony-track-memory-usage.md [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+Later, this function can be used to figure out the total memory used by this actor by using one of the following
+
+```
+    let size = @ponyint_actor_total_mem_size(this)
+    env.out.print("my size = " + size.string())
+```
+or,
+
+```
```
Comment (SeanTAllen) [inline content/reference/pony-track-memory-usage.md:35]:
```suggestion
```pony
```

File: content/reference/pony-track-memory-usage.md [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+
+### Building ponyc to track memory usage
+Tracking memory is a costly affair and is not enabled by default. Pony runtime needs to be built separately to track memory usage. 
+
+This is a two step process
+
+-  Configuring Pony
+   We need to ensure that at the time of configure step, `-DPONY_USE_MEMTRACK=true -DPONY_USE_MEMTRACK_MESSAGES=true` are passed to the cmake at the configure step.
```
Comment (SeanTAllen) [inline content/reference/pony-track-memory-usage.md:49]: There should be an empty line between 48 and 49 here.

File: content/reference/pony-track-memory-usage.md [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+   - For windows, this implies adding the `-DPONY_USE_MEMTRACK=true -DPONY_USE_MEMTRACK_MESSAGES=true` in the [configure](https://github.com/ponylang/ponyc/blob/make.ps1) section of make.ps1
+   - For Makefile based builds, passing appropriate arguments to the make [invocation](https://github.com/ponylang/ponyc/blob/Makefile)
+
+   There after, we run [configure command as usual](https://github.com/ponylang/ponyc/blob/main/BUILD.md)
+
+- Build
+  After the `configure` step we build pony from source as per the platform. We get `ponyc` enabled with tracking memory in the folder `build/release-memtrack_messages` folder.
```
Comment (SeanTAllen) [inline content/reference/pony-track-memory-usage.md:56]: empty line between 55 and 56 please.

**Comment by SeanTAllen** [comment]
I have no idea why the netlify preview isnt building for this.

File: content/reference/pony-track-memory-usage.md (hunk 3 above)
Comment (kapilash) [reply content/reference/pony-track-memory-usage.md]: moved.

**Comment by SeanTAllen** [comment]
If you feel comfortable with the process, can you rebase this against the latest `main` to pick up the spell checking and markdown linting that has been added?

**Comment by kapilash** [comment]
> Also, is this ready for another review?

Now please. Getting past the linter wasn't easy though.

**Comment by SeanTAllen** [comment]
This is an excellent start for this @kapilash. Thank you. I imagine as others try using these instructions. we'll expand them.

**Pull request merged**